File: leetcode/0937-reorder-data-in-log-files/solution.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cctype>
using namespace std;

class Solution {
public:
    vector<string> reorderLogFiles(vector<string>& logs) {
        vector<string> letterLogs, digitLogs;

        for (auto& log : logs) {
            size_t spaceIdx = log.find(' ');
            if (isdigit(log[spaceIdx + 1])) {
                digitLogs.push_back(log);
            } else {
                letterLogs.push_back(log);
            }
        }

        stable_sort(letterLogs.begin(), letterLogs.end(), [](const string& a, const string& b) {
            size_t aSpace = a.find(' ');
            size_t bSpace = b.find(' ');
            string aId = a.substr(0, aSpace), aRest = a.substr(aSpace + 1);
            string bId = b.substr(0, bSpace), bRest = b.substr(bSpace + 1);

            if (aRest != bRest) return aRest < bRest;
            return aId < bId;
        });

        letterLogs.insert(letterLogs.end(), digitLogs.begin(), digitLogs.end());
        return letterLogs;
    }
};

File: leetcode/0937-reorder-data-in-log-files/Solution.java
import java.util.*;

class Solution {
    public String[] reorderLogFiles(String[] logs) {
        List<String> letterLogs = new ArrayList<>();
        List<String> digitLogs = new ArrayList<>();

        for (String log : logs) {
            int spaceIdx = log.indexOf(' ');
            char firstContentChar = log.charAt(spaceIdx + 1);
            if (Character.isDigit(firstContentChar)) {
                digitLogs.add(log);
            } else {
                letterLogs.add(log);
            }
        }

        letterLogs.sort((a, b) -> {
            int aSpace = a.indexOf(' ');
            int bSpace = b.indexOf(' ');
            String aId = a.substring(0, aSpace), aRest = a.substring(aSpace + 1);
            String bId = b.substring(0, bSpace), bRest = b.substring(bSpace + 1);

            int cmp = aRest.compareTo(bRest);
            if (cmp != 0) return cmp;
            return aId.compareTo(bId);
        });

        letterLogs.addAll(digitLogs);
        return letterLogs.toArray(new String[0]);
    }
}
